client/src/standalone/player/player.ts

```diff
@@ -0,0 +1,182 @@
+     * between multiple PeerTube embeds. The string passed here must match the 
+     * `scope=` query parameter on the embed URL.
+     * 
+     * @param embedElement 
+     * @param scope 
+     */
+    constructor(
+        private embedElement : HTMLIFrameElement, 
+        private scope? : string
+    ) {
+        this.eventRegistrar.registerTypes(PASSTHROUGH_EVENTS)
+
+        this.constructChannel()
+        this.prepareToBeReady()
+    }
+
+    private eventRegistrar : EventRegistrar = new EventRegistrar()
+    private channel : Channel.MessagingChannel
+    private currentPosition = 0
+
+    /**
+     * Destroy the player object and remove the associated player from the DOM.
+     */
+    public destroy() {
+        this.embedElement.remove()
+    }
+
+    /**
+     * Listen to an event emitted by this player.
+     * 
+     * @param event One of the supported event types
+     * @param handler A handler which will be passed an event object (or undefined if no event object is included)
+     */
+    public addEventListener(event : PlayerEventType, handler : EventHandler<any>): boolean {
+        return this.eventRegistrar.addListener(event, handler)
+    }
+
+    /**
+     * Remove an event listener previously added with addEventListener().
+     * 
+     * @param event The name of the event previously listened to
+     * @param handler 
+     */
+    public removeEventListener(event : PlayerEventType, handler : EventHandler<any>): boolean {
+        return this.eventRegistrar.removeListener(event, handler)
+    }
+    
+    private constructChannel() {
+        this.channel = Channel.build({
+            window: this.embedElement.contentWindow,
+            origin: '*',
+            scope: this.scope || 'peertube'
+        })
+        this.eventRegistrar.bindToChannel(this.channel)
+    }
+ 
+    private _readyPromise : Promise<void>
+    private prepareToBeReady() {
+        let readyResolve, readyReject
+        this._readyPromise = new Promise<void>((res, rej) => {
+            readyResolve = res
+            readyReject = rej
+        })
+        
+        this.channel.bind('ready', success => success ? readyResolve() : readyReject())
+        this.channel.call({
+            method: 'isReady',
+            success: isReady => isReady ? readyResolve() : null
+        })
+    }
+
+    /**
+     * Promise resolves when the player is ready.
+     */
+    public get ready(): Promise<void> {
+        return this._readyPromise
+    }
+
+    private sendMessage<TIn, TOut>(method : string, params? : TIn): Promise<TOut> {
+        return new Promise<TOut>((resolve, reject) => {
+            this.channel.call({
+                method, params,
+                success: result => resolve(result),
+                error: error => reject(error)
+            })
+        })
+    }
+
+    /**
+     * Tell the embed to start/resume playback
+     */
+    async play() {
+        await this.sendMessage('play')
```

Comment:
```
play () {
  return this.sendMessage('play')
}
```

Reply:
I think it is better to always await instead of silently passing the promise back. This prevents a later modification like:

```
play() {
   try {
      return this.sendMessage('play');
   } catch (e) {
      // handle exception, but this will never ever fire
   }
}

Reply:
don't know why you would do such thing, we only `try` when we know the method implementation or when we do an `await` inside the block, but whatever it's not very important

Reply:
Say for instance we decide we need to throw an exception out of the `play()` that is inside the iframe. It triggers the `error` handler on JS channel and we throw it--- but say we want to have a proper error class to represent said type. But the object coming out of the channel will always be a plain object. So perhaps we do something like:

```
play() {
   try {
      return this.sendMessage('play');
   } catch (e) {
      if (e.type == 'unplayable') {
           throw new UnplayableError(e);
      }

      throw e;
   }
}
```

Now, the dev has made the mistake I am talking about with a concrete use case. Does this make more sense?

Reply:
:+1: